### PR TITLE
Support ResType on PropDefault annotation

### PR DESCRIPTION
**Pull request by pavlospt**
Initial work on adding support for `ResType` on `@PropDefault` ( #81 )

I went with the solution of having the `Builder` initializing the Props that have a default value marked as ResType, instead of passing down the `ComponentContext` which I tried to implement but it was introducing a quite huge change.

In order to reduce the review scope I have only added generation for `resolveDimenSizeRes()` and will add the rest before closing the PR, in order to have a naming and feature compliance agreement after the initial review. Also, I believe that after resolving the resource on a marked `@PropDefault` we should generate resource helper methods like we do for `@Prop`.

Let me know what you think! 

/cc: @emilsjolander 

P.S.: There is a big diff on the `ComponentImplGenerator` due to different Code Style, so If you could share the one you are using so that we could be aligned and reduce the diffs, would be great.

### CODE REVIEW

**Comment by emilsjolander** [comment]
@pavlospt Thanks! I'll have a look today or tomorrow and give more detailed feedback. Before that could you please revert the style changes you made? We have our preferred formatting style and would rather not change it, also makes the PR harder to review.

**Comment by pavlospt** [comment]
@emilsjolander Yeap that's what I wrote on the ps, could you share your formatting style so that I can revert it?  Because right now I don't see how I could that. Let me know if you have any ideas!

**Comment by pavlospt** [comment]
Code style issues are fixed!

**Comment by emilsjolander** [comment]
We don't have a formal formatting style really. We should probably work on one though 👍  Just look at how the rest of the code in the files are formatted and try to follow the same style.

File: sample/src/main/java/com/facebook/samples/litho/lithography/FeedItemCardSpec.java [hunk 1]
```diff
@@ -8,8 +8,10 @@
 
 package com.facebook.samples.litho.lithography;
 
+import android.graphics.drawable.Drawable;
```
Comment (emilsjolander) [inline sample/src/main/java/com/facebook/samples/litho/lithography/FeedItemCardSpec.java]: Seems unused

File: sample/src/main/java/com/facebook/samples/litho/lithography/FeedItemCardSpec.java [hunk 2]
```diff
@@ -8,8 +8,10 @@
...
 import com.facebook.litho.Column;
 
+import com.facebook.litho.annotations.ResType;
```
Comment (emilsjolander) [inline sample/src/main/java/com/facebook/samples/litho/lithography/FeedItemCardSpec.java]: seems unused

File: sample/src/main/java/com/facebook/samples/litho/playground/PlaygroundComponentSpec.java [hunk 3]
```diff
@@ -21,13 +23,16 @@
 
 import com.facebook.litho.annotations.LayoutSpec;
 import com.facebook.litho.annotations.OnCreateLayout;
-import com.facebook.litho.widget.Text;
 
 @LayoutSpec
 public class PlaygroundComponentSpec {
 
+  @PropDefault(resType = ResType.DIMEN_SIZE)
```
Comment (emilsjolander) [inline sample/src/main/java/com/facebook/samples/litho/playground/PlaygroundComponentSpec.java]: Could you revert these changes? Good to see example used but we don't want to add a bunch of stuff to the playground.

File: sample/src/main/res/values/dimens.xml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (emilsjolander) [inline sample/src/main/res/values/dimens.xml]: Remove this file.

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java [hunk 5]
```diff
@@ -733,4 +768,21 @@ private static MethodSpec generateReleaseMethod(SpecModel specModel) {
         .addStatement("$L.release(this)", BUILDER_POOL_FIELD)
         .build();
   }
+
+  private static String generateInitializer(PropDefaultModel propDefaultModel,
```
Comment (emilsjolander) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: If a line exceeds 100 we put all the parameters on the next line, including the first one 👍

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java [hunk 6]
```diff
@@ -733,4 +768,21 @@ private static MethodSpec generateReleaseMethod(SpecModel specModel) {
...
+      SpecModel specModel) {
+    final String defaultInitializer = "%s.%s";
+    StringBuilder builtInitializer = new StringBuilder();
+
+    if (propDefaultModel.hasResType()) {
+      if (propDefaultModel.mResType == ResType.DIMEN_SIZE) {
```
Comment (emilsjolander) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: We should handle all res types. See how this is done for the builder methods.

**Comment by pavlospt** [comment]
@emilsjolander thank you for the review, will try to fix them and finalize it.

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java (hunk 5 above)
Comment (pavlospt) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: @emilsjolander Do you mean 100 chars? This line seems to be 78 chars including the leading space.

**Comment by pavlospt** [comment]
@emilsjolander Fixed most of the review changes, one thing that is not clear though: in order to generate the initializers, should I create new variables? 

What I mean is: 

* If someone marks a `@PropDefault` with a `ResType` then it would most probably require it to be an `int` in order to accept the `R` reference, which means that we would need to generate a variable for the resolved type of the `ResType`.

**Comment by pavlospt** [comment]
I had a discussion with @charbgr for possible approaches on the above issue: 

The outcome was that we could add another key on `@PropDefault` called `resId` possibly and there the user would declare the `resourceId` that we would resolve. 

That means that the user gets to declare the same type of variable as the `@Prop` annotated and we would not need to create other variables as mediators.

**Comment by emilsjolander** [comment]
Looks pretty good! thanks 👍  Could you please add tests for the added features?

**Comment by pavlospt** [comment]
@emilsjolander Sure thing, is there a specific directory for the Annotation Processor tests ?

**Comment by passy** [comment]
Oh shoot, we're still working on getting the annotation processor tests open-sourced. There were some issues with Gradle and @IanChilds is working on fixing those.

**Comment by pavlospt** [comment]
@passy Hmmm, is there any way around this then ?

**Comment by IanChilds** [comment]
We should have them available today or tomorrow :)

**Comment by pavlospt** [comment]
@IanChilds Great!

**Comment by nscoding** [comment]
Any updates here? would be great to see this merged.

**Comment by pavlospt** [comment]
@nscoding Did not find some free time to write the tests yet..Hopefully will be able to do so, in the following week!

**Comment by pavlospt** [comment]
@IanChilds can you please review the last commit? I had to use `12345` as a constant for the `BuildGeneratorTest` additions due to limitation annotation fields that need to be a constant expression. Inside a ComponentSpec it works like a charm because the R field is resolved into a constant ID in advance.

File: .gitignore [hunk 7]
```diff
@@ -18,4 +18,9 @@ docs/.bundle/
 # Android Studio/IntelliJ
 .idea
 *.iml
+.gradle/
+lib/fb/
```
Comment (passy) [inline .gitignore]: We need those. :)
Can you revert that?

File: .gitignore [hunk 8]
```diff
@@ -18,4 +18,9 @@ docs/.bundle/
...
+lib/yoga/
+lib/yogajni/
+*/**.iml
```
Comment (passy) [inline .gitignore]: This should already be covered by the one above.

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java [hunk 9]
```diff
@@ -199,8 +216,26 @@ static TypeSpecDataHolder generateBuilder(SpecModel specModel) {
...
 
+    MethodSpec.Builder initResTypePropDefaultsSpec = null;
+
+    if (isResResolvable) {
+      initResTypePropDefaultsSpec = MethodSpec.methodBuilder("initPropDefaults");
+
+      for (PropDefaultModel propDefault : propDefaults) {
+        if (!propDefault.isResResolvable()) continue;
+
+        initResTypePropDefaultsSpec.addStatement("this.$L.$L = $L",
```
Comment (IanChilds) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: one param per line please :)

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java [hunk 10]
```diff
@@ -199,8 +216,26 @@ static TypeSpecDataHolder generateBuilder(SpecModel specModel) {
...
+            getImplMemberInstanceName(specModel), propDefault.getName(),
+            generatePropsDefaultInitializers(specModel, propDefault));
+      }
+    }
+
     propsBuilderClassBuilder.addMethod(initMethodSpec.build());
 
+    if (initResTypePropDefaultsSpec != null) {
```
Comment (IanChilds) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: why don't you just do everything at this point and then you don't have to do the null checks?

File: litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropDefaultModel.java [hunk 11]
```diff
@@ -53,6 +96,8 @@ public int hashCode() {
     int result = mType.hashCode();
     result = 17 * result + mName.hashCode();
     result = 31 * result + mModifiers.hashCode();
+    result = 31 * result + mResType.hashCode();
```
Comment (IanChilds) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropDefaultModel.java]: use a different number here and on the line below

File: litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropModel.java [hunk 12]
```diff
@@ -97,6 +97,17 @@ public boolean hasDefault(ImmutableList<PropDefaultModel> propDefaults) {
     return false;
   }
 
+  public PropDefaultModel getDefault(ImmutableList<PropDefaultModel> propDefaults) {
```
Comment (IanChilds) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropModel.java]: what is this for?

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java (hunk 10 above)
Comment (pavlospt) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: @IanChilds can you explain this a little bit more? :)

File: litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropModel.java (hunk 12 above)
Comment (pavlospt) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropModel.java]: @IanChilds unused, I removed it. Thanks!

**Comment by lucasr** [comment]
I'd update the docs for PropDefaults (https://fblitho.com/docs/props#prop-defaults) to cover this new feature :-)

**Comment by pavlospt** [comment]
@lucasr I can do this as well and include it in this PR :)

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java (hunk 9 above)
Comment (pavlospt) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: Done!

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java (hunk 10 above)
Comment (ahmedre) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: i think what he's saying is if you move this up to the bottom of the if on line 221, you can drop the null check.

File: litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java [hunk 13]
```diff
@@ -199,8 +216,26 @@ static TypeSpecDataHolder generateBuilder(SpecModel specModel) {
       initMethodSpec.addStatement("mRequired.clear()");
     }
 
+    MethodSpec.Builder initResTypePropDefaultsSpec = null;
```
Comment (ahmedre) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/generator/BuilderGenerator.java]: move into the if statement below

File: sample/src/main/java/com/facebook/samples/litho/playground/PlaygroundComponentSpec.java [hunk 14]
```diff
@@ -10,9 +10,7 @@
 
 import com.facebook.litho.Column;
 
-import com.facebook.yoga.YogaAlign;
-
-import com.facebook.yoga.YogaFlexDirection;
+import com.facebook.litho.widget.Text;
```
Comment (ahmedre) [inline sample/src/main/java/com/facebook/samples/litho/playground/PlaygroundComponentSpec.java]: nitpick: can you undo this change (it's nicer when we keep them alphabetized).

File: litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropDefaultModel.java [hunk 15]
```diff
@@ -53,6 +96,8 @@ public int hashCode() {
     int result = mType.hashCode();
     result = 17 * result + mName.hashCode();
     result = 31 * result + mModifiers.hashCode();
+    result = 43 * result + mResType.hashCode();
+    result = 65 * result + mResId;
```
Comment (ahmedre) [inline litho-processor/src/main/java/com/facebook/litho/specmodels/model/PropDefaultModel.java]: nitpick - for consistency, can you replace 65 with 47 or 53 or 59 or 61 to keep it as prime numbers?

**Comment by pavlospt** [comment]
@ahmedre Updated to reflect review requests. Can you check again :) ?

**Comment by ahmedre** [comment]
@pavlospt may you also update [props.md](https://github.com/facebook/litho/blob/master/docs/_docs/props.md) with these changes? btw, thanks for doing this!

lgtm otherwise, but will leave the remaining review checks to the Litho core team

**Comment by pavlospt** [comment]
@ahmedre Done :)
